src/xmipp/libraries/reconstruction/tomo_filter_coordinates.cpp

```diff
@@ -0,0 +1,248 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+ * 02111-1307  USA
+ *
+ *  All comments concerning this program package may be sent to the
+ *  e-mail address 'xmipp@cnb.csic.es'
+ ***************************************************************************/
+
+#include "tomo_filter_coordinates.h"
+#include <core/metadata_extension.h>
+#include <core/metadata_vec.h>
+#include <chrono>
+
+
+// --------------------------- INFO functions ----------------------------
+
+void ProgTomoFilterCoordinates::readParams()
+{
+	fnInTomo = getParam("--inTomo");
+	fnMask = getParam("--mask");
+	fnInCoord = getParam("--coordinates");
+	radius = getIntParam("--radius");
+    checkResThr = checkParam("--threshold");
+   	fnOut = getParam("-o");
+}
+
+
+void ProgTomoFilterCoordinates::defineParams()
+{
+	addUsageLine("This program filters a set of coordinates with one of two criteria:.");
+	addUsageLine("\nUsing a mask. In this case a set of coordinates and a mask with the regions to preserve ");
+    addUsageLine("coordinates different than 0 need to be input. If only these two options are input this ");
+    addUsageLine("criteria will be applied, if not the second one will be applied");
+	addUsageLine("\nUsing a resolution. In this case a set of coordinates, a resolution map, and a resolution ");
+    addUsageLine("percentile to select the number of coordinates to be saved after the scoring need o be ");
+    addUsageLine("input. If these three options are input then this criteria will be applied.");
+	addParamsLine("  --inTomo <mrcs_file=\"\">                                : Input volume (mask or resolution map).");
+	addParamsLine("  [--mask <xmd_file=\"\">]                               : Input xmd file containing the 3D coordinates.");
+	addParamsLine("  --coordinates <xmd_file=\"\">                               : Percentile resolution threshold.");
+	addParamsLine("  --radius <radius=50>                               : Radius of the neighbourhood of the coordinates to get resolution score.");
+    addParamsLine("  [--threshold <outCoord=\"filteredCoordinates3D.xmd\">]   : Output file containing the filtered 3D coordinates.");
+	addParamsLine("  -o <outCoord=\"filteredCoordinates3D.xmd\">   : Output file containing the filtered 3D coordinates.");
+}
+
+
+// --------------------------- HEAD functions ----------------------------
+
+void ProgTomoFilterCoordinates::filterCoordinatesWithMask(MultidimArray<int> &inputVolume)
+{
+	Point3D<int> coord3D;
+	for (size_t i = 0; i < inputCoords.size(); i++)
+	{
+		coord3D = inputCoords[i];
+
+		if(coord3D.z < 0 || coord3D.z > (zDim-1) || coord3D.y < 0 || coord3D.y > (yDim-1) || coord3D.x < 0 || coord3D.x > (xDim-1))
+		{
+			std::cout << "WARNNING: Coordinate at (x=" << coord3D.x<< ", y=" << coord3D.y << ", z=" << coord3D.z << ") erased due to its out of the mask." << std::endl;
+			inputCoords.erase(inputCoords.begin()+i);
```

Comment:
I think that removing item from vector over which you iterate is undefined operation